### DISCUSSION TITLE
Disable instant menu trigger when on end of line.

### DISCUSSION THREAD

**Discussion by redoxahmii** [saghen/blink.cmp#566] [Q&A]
Whenever i use `i` or `a` to enter insert mode at the end of the line the menu appears instantly causing the item to be accepted instead of being able to go a line below by pressing `<CR>`.
Anyway to solve this ?
And how are the items in the menu being sorted as it prioritizes Text in the buffer more than the actual responses from LSP or when using `path` cmp.
Any help for that is appreciated too.
I've tried using `fuzzy.use_typo_resistance` but that does not seem to be the solution for me.
At this point i would rather have the whole buffer text suggestion turned off.

**Comment by redoxahmii** [comment]
the cmp menu should trigger only when i actually start typing instead of whenever i enter insert mode is what i want to figure out

**Comment by mnchaser** [comment]
I am having the same issue.  Here's some additional context (and a thank you in advance).  

At the end of a line, pressing "a" will activate the cmp window and default to the first suggestion in the list.  You can't press Enter to simply create a new line, you will accept the cmp suggestion. Pressing Esc will reset the cursor/mode, so that isn't a workaround.  Space bar will dismiss the window, but then you get a trailing space on that line.

Steps to reproduce (LazyVim): 
1. In normal mode, place cursor at and of a line
2. press "a" to enter insert mode (press nothing else)
3. auto-complete box will appear with a word pre-selected 

-The same situation applies if you already have a trailing space at the end of your line and you try to insert ("i") before the space. 
-The same situation applies if you have <style></style> and you enter insert mode in between those elements.

Possibly related: 
-The same situation applies if you are typing <style  .... the moment you press ">" to close the element, </style> closing element is inserted/auto-completed (which is good), but then the cmp window pops up. Can't press Esc or Enter, only Space (same workaround as above).  

I have tried:
trigger.prefetch_on_insert = false,
trigger.show_on_insert_on_trigger_character = false,
     and
list.selection="manual"

but none of these seem to make a difference.  Thank you!

![blink issue](https://github.com/user-attachments/assets/2c890df9-b677-49d5-88b1-7ea0cb4df911)

**Reply by redoxahmii** [reply]
Thank you for adding to it further.
I am on LazyVim as well and using `list.selection = "auto_insert"` seems to be something closer but this requires you to use `<C-n>` or `<C-p>` to select that selection and you iterate further on it.
This has its own set of issues as when you need to insert a snippet it does not function properly in that state and then requires you to first select an entry and then press `<cr>` to run the snippet.

There should be a good default to how `nvim-cmp` works so it's easier for people to adapt.

**Reply by mnchaser** [reply]
I've also been playing with (trying to) block additional characters from being triggers, via: 
```
show_on_x_blocked_trigger_characters = { "'", '"', "(", ")", "<", ">", "{", "}", ",", "[", "]", "/" },
```
(which is a confusing setting name, but I believe is a list of blocked trigger characters.)  
But, I can't get it to work.  The default ones (first three in that list)  seem to work  ' " (  do NOT trigger the cmp window when you insert after them.   
I have a blink.lua file in my plugins folder like every other plugin settings I've modified, but it doesn't look to be taking any of the settings I am putting in there.

**Reply by redoxahmii** [reply]
This is the config i have right now which makes it closer to how it works in `nvim-cmp` for me.
```lua
{
    "saghen/blink.cmp",
    ---@module 'blink.cmp'
    ---@type blink.cmp.Config
    opts = {
      completion = {
        menu = {
          draw = {
            columns = { { "label", "label_description", gap = 1 }, { "kind_icon", "kind" } },
          },
        },
        list = {
          selection = "preselect",
        },
        trigger = {
          show_on_insert_on_trigger_character = false,
        },
      },
    },
  }
  ```
i missed the `opts.trigger` table somehow but looking into it this at least makes it close to how `nvim-cmp` handles it and also is not causing those unnecessary triggers.
Thank you for mentioning `opts.trigger`!
Do tell me if this works for you or there is still some issues for you.

**Reply by redoxahmii** [reply]
One more issue that I've with `blink.cmp` right now is the prioritization of what is shown as the first entry which `nvim-cmp` handled quite well.
Some LSPs like `emmet-language-server` entries are not even showing in the menu which I've no clue atm.

**Reply by redoxahmii** [reply]
The options you mentioned does not seem to work for me as well specifying anything there doesn't do anything the only way it is working for me is using the `trigger.show_on_insert_on_trigger_character = false` so it does not open that window whenever inserting.
From what I've understood is that symbols like `>` are somehow being classified as trigger words.
Some LSPs provide these as well but it mentions in the docs that it is disabled by default to not use LSPs triggers.

The only annoyance left right now is whenever you open any bracket it gets automatically triggered which for the life of me i can't find a way to turn off.

**Reply by redoxahmii** [reply]
so i changed `trigger.show_on_trigger_character = false` and reverted `trigger.show_on_insert_on_trigger_character` to the default true and now it is not triggering the menu whenever i open a bracket so my suspicion of that was correct that those are being considered as trigger words.

**Reply by mnchaser** [reply]
FWIW - I noticed I didn't have completions = {} in my opts plugin config. When I added this in, the config seems to be working.  Also, like you, it's now not popping up auto-completion at the end of { or > anymore.... so for now, something in my config appears to be working and I need to spend some time backing things out to see which it is.

**Reply by redoxahmii** [reply]
The issues i was having with it triggering on `>` have been resolved and added by default now which might be why it's working properly for you now.
I did explicitly set the options but they are now set by default and most of these issues are resolved.
